spikeinterface/core/waveform_tools.py

```diff
@@ -0,0 +1,184 @@
+"""
+This module contain low level function to retrieve snippet of traces aka "spike waveforms".
+
+This is internally used by WaveformExtractor but also can be used as a sorting component.
+
+It is a 2 steps way :
+  1. allocate buffer (file or memory)
+  2. extract and distribute snipet to buffer (in parallel optionaly)
+
+
+"""
+import platform
+
+import numpy as np
+
+from .job_tools import ChunkRecordingExecutor, ensure_n_jobs, _shared_job_kwargs_doc
+from .core_tools import make_shared_array
+
+
+def allocate_waveforms(recording, spikes, unit_ids, nbefore, nafter, mode='memmap', folder=None, dtype=None, sparsity_mask=None):
+    """
+    
+    
+    sparsity_mask:
+        If not None must
+        shape=(len(unit_ids), len(channel_ids)
+        dtype='bool'
+        
+    
+    """
+    
+    nsamples = nbefore + nafter
+    
+    dtype = np.dtype(dtype)
+    if mode =='shared_memory':
+        assert folder is None
+    
+    # prepare buffers
+    wfs_arrays = {}
+    wfs_arrays_info = {}
+    for unit_ind, unit_id in enumerate(unit_ids):
+        n_spikes = np.sum(spikes['unit_ind'] == unit_ind)
+        if sparsity_mask is None:
+            num_chans = recording.get_num_channels()
+        else:
+            num_chans = np.sum(sparsity_mask[unit_ind, :])
+        shape = (n_spikes, nsamples, num_chans)
+        
+        if mode =='memmap':
+            filename = folder / f'waveforms_{unit_id}.npy'
+            arr = np.lib.format.open_memmap(filename, mode='w+', dtype=dtype, shape=shape)
```

Comment:
@samuelgarcia probably raw strings would fix this for windows
```suggestion
            arr = np.lib.format.open_memmap(rf"{str(filename)}", mode='w+', dtype=dtype, shape=shape)
```